### PR TITLE
[core] Annotations refactor

### LINKED ISSUE #1488: Switching styles turns annotations invisible
Changing the style causes any annotations in view to become invisible. Any popover remains visible, and invisible annotations continue to respond to taps.

pindrop

### LINKED ISSUE #1675: No annotations at initial zoom level
If you follow the “[First steps with Mapbox GL for iOS](https://www.mapbox.com/guides/first-steps-gl-ios/#adding-a-marker)” guide, you wind up initializing an `MGLMapView`, centering it on the U.S. President’s front yard, and placing an annotation there, all within `-[ViewController viewDidLoad]`. But if you do all that in `-viewDidLoad`, the annotations fail to display at the initial zoom level. They appear when you zoom in or out to a different zoom level and disappear when you reenter the initial zoom level.

At launch, the map view happens to show z15 tiles:

initial

Then zoom in, right to the cusp of z16:

in

And zoom in just a little farther to z17:

inner

Or zoom out, right to the cusp of z14:

out

And a little further out to 14:

outer

The upshot is that our guide does not work as designed. We didn’t catch this issue in development because the iOS demo app only adds annotations in response to an action sheet button, not within `-viewDidLoad`.

Reproduces in ios-v0.3.1 as well as master, regardless of style.

/cc @incanus

### LINKED ISSUE #2095: Markers don't show up intil map is zoomed or op is delayed
There's a problem in Android where Markers don't show up until i zoom the map. It does work if i add markers with a timer Handler with 1 sec delay to it.

### LINKED ISSUE #2322: Simple polylines & polygons don't appear if initialized with map
Polylines and polygons that don't require much processing before being added to the map fail to ever appear, if initialized concurrently with the map. Adding the slightest delay or drawing the shapes in `viewDidAppear` causes them to appear as expected.

The [GeoJSON example](https://www.mapbox.com/ios-sdk/examples/line-geojson/) works when drawn directly in `viewDidLoad`.

#2289 fixed this problem for markers ~~and complex shapes~~, but I clearly must have missed the simple case during testing.

Demo ([based on this](https://www.mapbox.com/ios-sdk/examples/polygon/)):

```swift
import Mapbox

class ViewController: UIViewController, MGLMapViewDelegate {
    
    var mapView: MGLMapView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        mapView = MGLMapView(frame: view.bounds)
        mapView.autoresizingMask = .FlexibleWidth | .FlexibleHeight
        
        mapView.setCenterCoordinate(CLLocationCoordinate2D(latitude: 45.520486,
                                                           longitude: -122.673541),
                                    zoomLevel: 11, animated: false)
        view.addSubview(mapView)
        
        mapView.delegate = self
        
        // this works
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0), dispatch_get_main_queue(), {
            self.drawShape()
        })

        // this does not work
        drawShape()
    }
    
    override func viewDidAppear(animated: Bool) {
        // this works
        drawShape()
    }

    func drawShape() {
        // Create a coordinates array to hold all of the coordinates for our shape.
        var coordinates = [
            CLLocationCoordinate2D(latitude: 45.522585, longitude: -122.685699),
            CLLocationCoordinate2D(latitude: 45.534611, longitude: -122.708873),
            CLLocationCoordinate2D(latitude: 45.530883, longitude: -122.678833),
            CLLocationCoordinate2D(latitude: 45.547115, longitude: -122.667503),
            CLLocationCoordinate2D(latitude: 45.530643, longitude: -122.660121),
            CLLocationCoordinate2D(latitude: 45.533529, longitude: -122.636260),
            CLLocationCoordinate2D(latitude: 45.521743, longitude: -122.659091),
            CLLocationCoordinate2D(latitude: 45.510677, longitude: -122.648792),
            CLLocationCoordinate2D(latitude: 45.515008, longitude: -122.664070),
            CLLocationCoordinate2D(latitude: 45.502496, longitude: -122.669048),
            CLLocationCoordinate2D(latitude: 45.515369, longitude: -122.678489),
            CLLocationCoordinate2D(latitude: 45.506346, longitude: -122.702007),
            CLLocationCoordinate2D(latitude: 45.522585, longitude: -122.685699),
        ]
        
        let shape = MGLPolygon(coordinates: &coordinates, count: UInt(coordinates.count))
        
        mapView.addAnnotation(shape)
    }
}
```

/cc @incanus

### PR DESCRIPTION
**Pull request by jfirebaugh**
Significant refactoring of annotations. Annotation-related code that was in `MapContext` and `StyleParser` is now in `AnnotationManager`.

Includes additional tests for line and fill annotations.

Cumulates in rewriting the annotation invalidation strategy:

* First, move style mutation code out of StyleParser and into AnnotationManager, coalescing it with the mutation code for shape layers.
* Second, allow AnnotationManager to keep track of stale tiles entirely internally; avoid passing sets of TileIDs around.
* Third, correct the logic for invalidating the shape source. Since AnnotationManager does not track shape invalidations on a tile-by-tile basis, don't try to invalidate the shape source tile-by-tile.

Fixes #1675 - @picciano I used code derived from your test app to confirm. If you have time to confirm on your app, that would be awesome.
Probably fixes #2322 - @friedbunny can you confirm?
Probably fixes #2095 - @erf can you confirm?
Might partially fix #1488 - I will add a test and confirm.

### CODE REVIEW

**Comment by picciano** [comment]
@jfirebaugh I will do that first thing tomorrow. thanks

**Comment by jfirebaugh** [comment]
Hmm, definitely seeing some issues with this after testing some more scenarios. Stand by.

**Comment by erf** [comment]
i running the app fails on the latest 'master' or 'annotations' branch, i am told 'couldn't find "libmapbox-gl.so"', even if i have built it using 'make android', i went back to 4b58760 and then it worked ok.
<pre>
09-25 10:36:45.520    1535-1535/? E/AndroidRuntime﹕ FATAL EXCEPTION: main
    Process: com.mapbox.mapboxgl.testapp, PID: 1535
    java.lang.RuntimeException: Unable to start activity ComponentInfo{com.mapbox.mapboxgl.testapp/com.mapbox.mapboxgl.testapp.MainActivity}: android.view.InflateException: Binary XML file line #33: Error inflating class com.mapbox.mapboxgl.views.MapView
            at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2357)
            at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2419)
            at android.app.ActivityThread.access$900(ActivityThread.java:154)
            at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1321)
            at android.os.Handler.dispatchMessage(Handler.java:102)
            at android.os.Looper.loop(Looper.java:135)
            at android.app.ActivityThread.main(ActivityThread.java:5293)
            at java.lang.reflect.Method.invoke(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:372)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:904)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:699)
     Caused by: android.view.InflateException: Binary XML file line #33: Error inflating class com.mapbox.mapboxgl.views.MapView
            at android.view.LayoutInflater.createView(LayoutInflater.java:633)
            at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:743)
            at android.view.LayoutInflater.rInflate(LayoutInflater.java:806)
            at android.view.LayoutInflater.rInflate(LayoutInflater.java:809)
            at android.view.LayoutInflater.rInflate(LayoutInflater.java:809)
            at android.view.LayoutInflater.rInflate(LayoutInflater.java:809)
            at android.view.LayoutInflater.inflate(LayoutInflater.java:504)
            at android.view.LayoutInflater.inflate(LayoutInflater.java:414)
            at android.view.LayoutInflater.inflate(LayoutInflater.java:365)
            at android.support.v7.app.AppCompatDelegateImplV7.setContentView(AppCompatDelegateImplV7.java:255)
            at android.support.v7.app.AppCompatActivity.setContentView(AppCompatActivity.java:109)
            at com.mapbox.mapboxgl.testapp.MainActivity.onCreate(MainActivity.java:84)
            at android.app.Activity.performCreate(Activity.java:5990)
            at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1106)
            at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2310)
            at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2419)
            at android.app.ActivityThread.access$900(ActivityThread.java:154)
            at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1321)
            at android.os.Handler.dispatchMessage(Handler.java:102)
            at android.os.Looper.loop(Looper.java:135)
            at android.app.ActivityThread.main(ActivityThread.java:5293)
            at java.lang.reflect.Method.invoke(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:372)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:904)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:699)
     Caused by: java.lang.reflect.InvocationTargetException
            at java.lang.reflect.Constructor.newInstance(Native Method)
            at java.lang.reflect.Constructor.newInstance(Constructor.java:288)
            at android.view.LayoutInflater.createView(LayoutInflater.java:607)
            at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:743)
            at android.view.LayoutInflater.rInflate(LayoutInflater.java:806)
            at android.view.LayoutInflater.rInflate(LayoutInflater.java:809)
            at android.view.LayoutInflater.rInflate(LayoutInflater.java:809)
            at android.view.LayoutInflater.rInflate(LayoutInflater.java:809)
            at android.view.LayoutInflater.inflate(LayoutInflater.java:504)
            at android.view.LayoutInflater.inflate(LayoutInflater.java:414)
            at android.view.LayoutInflater.inflate(LayoutInflater.java:365)
            at android.support.v7.app.AppCompatDelegateImplV7.setContentView(AppCompatDelegateImplV7.java:255)
            at android.support.v7.app.AppCompatActivity.setContentView(AppCompatActivity.java:109)
            at com.mapbox.mapboxgl.testapp.MainActivity.onCreate(MainActivity.java:84)
            at android.app.Activity.performCreate(Activity.java:5990)
            at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1106)
            at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2310)
            at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2419)
            at android.app.ActivityThread.access$900(ActivityThread.java:154)
            at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1321)
            at android.os.Handler.dispatchMessage(Handler.java:102)
            at android.os.Looper.loop(Looper.java:135)
            at android.app.ActivityThread.main(ActivityThread.java:5293)
            at java.lang.reflect.Method.invoke(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:372)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:904)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:699)
     Caused by: java.lang.UnsatisfiedLinkError: dalvik.system.PathClassLoader[DexPathList[[zip file "/data/app/com.mapbox.mapboxgl.testapp-1/base.apk"],nativeLibraryDirectories=[/vendor/lib, /system/lib]]] couldn't find "libmapbox-gl.so"
            at java.lang.Runtime.loadLibrary(Runtime.java:366)
            at java.lang.System.loadLibrary(System.java:988)
            at com.mapbox.mapboxgl.views.NativeMapView.<clinit>(NativeMapView.java:37)
            at com.mapbox.mapboxgl.views.MapView.initialize(MapView.java:364)
            at com.mapbox.mapboxgl.views.MapView.<init>(MapView.java:318)
            at java.lang.reflect.Constructor.newInstance(Native Method)
            at java.lang.reflect.Constructor.newInstance(Constructor.java:288)
            at android.view.LayoutInflater.createView(LayoutInflater.java:607)
            at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:743)
            at android.view.LayoutInflater.rInflate(LayoutInflater.java:806)
            at android.view.LayoutInflater.rInflate(LayoutInflater.java:809)
            at android.view.LayoutInflater.rInflate(LayoutInflater.java:809)
            at android.view.LayoutInflater.rInflate(LayoutInflater.java:809)
            at android.view.LayoutInflater.inflate(LayoutInflater.java:504)
            at android.view.LayoutInflater.inflate(LayoutInflater.java:414)
            at android.view.LayoutInflater.inflate(LayoutInflater.java:365)
            at android.support.v7.app.AppCompatDelegateImplV7.setContentView(AppCompatDelegateImplV7.java:255)
            at android.support.v7.app.AppCompatActivity.setContentView(AppCompatActivity.java:109)
            at com.mapbox.mapboxgl.testapp.MainActivity.onCreate(MainActivity.java:84)
            at android.app.Activity.performCreate(Activity.java:5990)
            at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1106)
            at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2310)
            at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2419)
            at android.app.ActivityThread.access$900(ActivityThread.java:154)
            at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1321)
            at android.os.Handler.dispatchMessage(Handler.java:102)
            at android.os.Looper.loop(Looper.java:135)
            at android.app.ActivityThread.main(ActivityThread.java:5293)
            at java.lang.reflect.Method.invoke(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:372)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:904)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:699)
</pre>

**Comment by jfirebaugh** [comment]
Ok, I don't think that Android issue is caused by these changes. 

I fixed a bug with style cascading that was causing shapes to appear black. Should be good to test now.

This does _not_ fix #1488 -- points are still invisible and there are messages like "Can't find sprite named 'com.mapbox.sprites.51'" in the output when you switch a style after adding point annotations.

**Comment by friedbunny** [comment]
As of e9ece31 this does fix #2322. :+1:

**Comment by erf** [comment]
I build and run ok on latest master 81fde42

File: src/mbgl/map/map_context.cpp [hunk 1]
```diff
@@ -420,7 +313,6 @@ void MapContext::onResourceLoadingFailed(std::exception_ptr error) {
 }
 
 void MapContext::onSpriteStoreLoaded() {
-    updateAnnotationTilesIfNeeded();
 }
```
Comment (incanus) [inline src/mbgl/map/map_context.cpp]: We don't necessarily need `onSpriteStoreLoaded()` anymore. This was added in https://github.com/mapbox/mapbox-gl-native/commit/0d899df9ddc9a7a9a8bfbac2c4b077ce26da64e9 to address #1675, but this PR obsoletes that.

**Comment by incanus** [comment]
> This does not fix #1488 -- points are still invisible and there are messages like "Can't find sprite named 'com.mapbox.sprites.51'" in the output when you switch a style after adding point annotations.

Is it possible to just send another `Update::Annotations` after style changes somehow to fix this?

**Comment by incanus** [comment]
Seeing odd staleness with the built-in 1,000 points test: 

anigif-1443465578

**Comment by incanus** [comment]
Same for 100 points, actually.

**Comment by jfirebaugh** [comment]
> Is it possible to just send another `Update::Annotations` after style changes somehow to fix this? 

No, fixing this requires some bigger changes: keeping a record of which sprite images were added via `Map::setSprite` and then re-add them to the new style. Or, better, removing `Map::setSprite` and associating the image directly with the annotation object.

**Comment by jfirebaugh** [comment]
> Seeing odd staleness with the built-in 1,000 points test

Found, fixed, added a test.

**Comment by picciano** [comment]
@jfirebaugh  This is working in the TallyGo app now. Thanks much.

**Comment by incanus** [comment]
> Found, fixed, added a test.

Where at @jfirebaugh? I'm only seeing https://github.com/mapbox/mapbox-gl-native/commit/0156ac7c8ce7d22b8a68cc37fcb03b1271ef7a4e as recent and not seeing how this works. 

But I am seeing the fix!

**Comment by jfirebaugh** [comment]
I messed up splitting `Source::invalidateTiles` into two methods, and one of them invalidated `tile_data` but not `tiles`. https://github.com/mapbox/mapbox-gl-native/blob/6d357eb0c87897ef99926a4b2b12d0ae3354e7d8/test/api/annotations.cpp#L84-84 will catch that.

**Pull request merged**